### PR TITLE
fix(lambda-event-sources): kafka event source expects credentials even when accessed via vpc

### PR DESCRIPTION
**Pull request by therockstorm**
Fixes https://github.com/aws/aws-cdk/issues/14038

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts:138]: update this to return `undefined` if the list is empty

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts:23]: Is there a way to detect this state and error? 
For instance, if the lambda is not attached to a vpc. Wouldn't this imply that the access will be over the internet?

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts:85]: Would prefer if you changed the test case `default` to not have a secret. And updated this test case to include the secret.

Same applies for the self managed kafka test suite.

**Review by nija-at — CHANGES_REQUESTED** [review]
Thanks for submitting this PR @therockstorm.

Please see my comments below.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts
Comment (therockstorm) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts:23]: Added an error on line 161.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts
Comment (therockstorm) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts]: Updated both MSK and self-managed.

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts
Comment (therockstorm) [reply packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts]: Flipped these around

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts:161]: Thanks for adding. Don't we need the same for managed kafka as well?

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts:140]: can you update the 'default' case to not use & assert `secret`? 

The 'default' test case is intended to assert that the expected defaults are configured. This means that they must only use all the required props.

**Review by nija-at — CHANGES_REQUESTED** [review]
Thanks for making the changes. See a couple of more comments below.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts
Comment (therockstorm) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts:161]: From https://www.amazonaws.cn/en/msk/faqs/,

>  Q: Does Amazon MSK run in an Amazon VPC?
> Yes, Amazon MSK always runs within an Amazon VPC managed by the Amazon MSK service.

Comment (therockstorm) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts:161]: I'm not sure if this means `secret` is only valid for self-managed Kafka or not, so I left it.

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts
Comment (therockstorm) [reply packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts:140]: This means `'default'` will throw an error since either `secret` or `vpc` is required. I've made the change.

**Comment by therockstorm** [comment]
This is ready for another look, @nija-at. Thanks!

**Comment by therockstorm** [comment]
Is there anything else you need from me, @nija-at? A project I'm working on is blocked by this fix so I'd really appreciate getting this in if possible.

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts
Comment (nija-at) [reply packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts:140]: The default test case must specify all the required properties and then verify that the other properties correctly configure the expected defaults. So you will need to change this to match up with that.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts
Comment (nija-at) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts:161]: Let's leave this as is for now. We can come back to it later when we have more information.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts
Comment (nija-at) [inline packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts:112]: please move this comment to the new location as well.

**Review by nija-at — CHANGES_REQUESTED** [review]
Hey @therockstorm. Thanks for the updates.

Sorry for the delay here. This seems to have slipped through my usual search filters. Now that it's assigned to me, it won't anymore.

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts
Comment (therockstorm) [reply packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts:140]: I'm not sure I understand what you're asking for. Do you mean "required" by TypeScript? If so, that's what this test is doing. `secret` and `vpc` are both optional parameters because only one or the other is required. If either is missing, I'm not aware of any default we could safely choose.

File: packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts
Comment (therockstorm) [reply packages/@aws-cdk/aws-lambda-event-sources/lib/kafka.ts]: Moved

File: packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts
Comment (nija-at) [reply packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts:140]: Let me try again and see if I can make it clearer.

The 'default' test case is testing that defaults for optional parameters are set as expected. In your case 'secret' or 'vpc' is required.

So I would suggest configuring one of them and then asserting that the rest of the optional parameters (excluding 'secret') is configured as expected.

The test case as you have right now is required, but it shouldn't be called the 'default' test case and is separate from it.

Is that clearer?

Comment (therockstorm) [reply packages/@aws-cdk/aws-lambda-event-sources/test/test.kafka.ts:140]: Thanks, that clears it up. I updated the `default` test.

**Review by nija-at — DISMISSED** [review]
Thanks for making the changes!

**Comment by therockstorm** [comment]
Sorry, I think I need another approval, @nija-at. I thought the branch had to be updated because the `validate-pr` step wasn't turning green, but it looks like doing that dismissed your approval. Do I need to do anything else or will bots take care of it from here?

**Comment by nija-at** [comment]
@therockstorm - got it. The bots should take care of doing everything from here on. Tag me on a comment here if the PR is not merged within the next day.

**Review by nija-at — APPROVED** [review]

**Comment by therockstorm** [comment]
@nija-at It's been over 24 hours since your approval and still no bot activity.

**Pull request merged**
